File: prismpath-telemetry-rs/src/zeckendorf.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Zeckendorf / Fibonacci codec — the self-framing wire for the decision-preserving telemetry stream.

fn fibs_upto(upper_bound: usize) -> Vec<usize> {
    // `fibs` is seeded with [1, 2] and only ever grows, so `.last()` is always Some.
    let mut fibs = vec![1, 2];
    while *fibs.last().expect("fibs seeded non-empty") <= upper_bound {
        let next = fibs[fibs.len() - 1] + fibs[fibs.len() - 2];
        fibs.push(next);
    }
    if *fibs.last().expect("fibs seeded non-empty") > upper_bound {
        fibs.pop();
    }
    fibs
}

/// Fibonacci code of a positive integer `n` (>= 1) as a bit-string ending in `"11"`.
pub fn encode(value: usize) -> Result<String, String> {
    if value < 1 {
        return Err(format!("Fibonacci coding is for positive integers; got {value}"));
    }
    let fibs = fibs_upto(value);
    let mut bits = vec!['0'; fibs.len()];
    let mut rem = value;
    for index in (0..fibs.len()).rev() {
        if fibs[index] <= rem {
            bits[index] = '1';
            rem -= fibs[index];
        }
    }
    assert_eq!(rem, 0, "Zeckendorf decomposition failed for {value}");
    let mut out: String = bits.into_iter().collect();
    out.push('1'); // append terminator -> unique trailing '11'
    Ok(out)
}

/// Inverse of `encode`: one Fibonacci code (bit-string ending in `"11"`) -> the integer.
pub fn decode(code: &str) -> Result<usize, String> {
    if code.len() < 2 || !code.ends_with("11") {
        return Err(format!("not a Fibonacci code (must end in '11'): {code:?}"));
    }
    let zeck = &code[..code.len() - 1]; // strip terminator '1'
    let mut fibs = vec![1, 2];
    while fibs.len() < zeck.len() {
        let next = fibs[fibs.len() - 1] + fibs[fibs.len() - 2];
        fibs.push(next);
    }
    let mut sum = 0;
    for (index, ch) in zeck.chars().enumerate() {
        if ch == '1' {
            sum += fibs[index];
        }
    }
    Ok(sum)
}

/// Concatenate the codes; each code's trailing `"11"` frames the next — zero header, self-delimiting.
/// `Err` if any value is `< 1` (Fibonacci coding is for positive integers; the wire sends symbol+1).
pub fn encode_stream(values: &[usize]) -> Result<String, String> {
    let mut out = String::new();
    for &value in values {
        out.push_str(&encode(value)?);
    }
    Ok(out)
}

/// Split a concatenated stream at each self-framing `"11"` and decode each code.
/// A trailing run of bits with no terminator is an incomplete final frame and is dropped.
pub fn decode_stream(bits: &str) -> Vec<usize> {
    let mut out = Vec::new();
    let chars: Vec<char> = bits.chars().collect();
    let bit_count = chars.len();
    let mut start = 0;
    let mut cursor = 0;
    while cursor < bit_count {
        if chars[cursor] == '1' && cursor + 1 < bit_count && chars[cursor + 1] == '1' {
            let slice: String = chars[start..=cursor + 1].iter().collect();
            if let Ok(val) = decode(&slice) {
                out.push(val);
            }
            cursor += 2;
            start = cursor;
        } else {
            cursor += 1;
        }
    }
    out
}

/// Strict variant for a receiving codec: the lenient `decode_stream` silently drops an incomplete
/// final frame, which is right for the self-heal/inspect paths but hides truncation from a decoder
/// that must surface corruption. Here any leftover bits after the last complete code are an error
/// unless they are all `'0'` — a run of zero bits is legitimate byte-alignment padding
/// (`packed::pack`), while a leftover `'1'` means a truncated or corrupted final code.
pub fn decode_stream_strict(bits: &str) -> Result<Vec<usize>, String> {
    let mut out = Vec::new();
    let chars: Vec<char> = bits.chars().collect();
    let bit_count = chars.len();
    let mut start = 0;
    let mut cursor = 0;
    while cursor < bit_count {
        if chars[cursor] == '1' && cursor + 1 < bit_count && chars[cursor + 1] == '1' {
            let slice: String = chars[start..=cursor + 1].iter().collect();
            out.push(decode(&slice)?);
            cursor += 2;
            start = cursor;
        } else {
            cursor += 1;
        }
    }
    if chars[start..].contains(&'1') {
        return Err(format!(
            "truncated Fibonacci stream: {} trailing bits contain a '1' with no terminator",
            bit_count - start
        ));
    }
    Ok(out)
}

File: prismpath/telemetry/zeckendorf.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Zeckendorf / Fibonacci codec — the self-framing wire for the decision-preserving telemetry stream.

Every positive integer has a unique Zeckendorf representation (a sum of non-consecutive Fibonacci
numbers), so its bit pattern never contains ``11``. The Fibonacci code appends a terminal ``1``, making a
trailing ``11`` that appears nowhere else — a permanent, data-agnostic codeword boundary. Variable-length,
prefix-free, ``O(log n)``, and densest on small integers (``1 -> 11``, ``2 -> 011``, ``3 -> 0011``,
``4 -> 1011``), which is exactly where delta-differenced telemetry lives.

Fibonacci coding is defined for POSITIVE integers (n >= 1). Callers that need to send 0 or signed deltas
offset/zig-zag at the symbol layer, not here — this module stays the pure, doc-faithful wire.

Reference (bit-string) implementation for Phase A: correctness + measurement first. The u64-accumulate
CPU path and the FPGA shift-register codec are Phase C.
"""
from __future__ import annotations

from typing import Iterable, List


def _fibs_upto(limit: int) -> List[int]:
    """Ascending Fibonacci basis ``[1, 2, 3, 5, 8, ...]`` (F2, F3, F4, ...) with the largest term <= n."""
    fibs = [1, 2]
    while fibs[-1] <= limit:
        fibs.append(fibs[-1] + fibs[-2])
    if fibs[-1] > limit:
        fibs.pop()
    return fibs


def encode(value: int) -> str:
    """Fibonacci code of a positive integer ``n`` (>= 1) as a bit-string ending in ``11``."""
    if value < 1:
        raise ValueError(f"Fibonacci coding is for positive integers; got {value}")
    fibs = _fibs_upto(value)
    bits = ["0"] * len(fibs)                 # bits[i] <-> fibs[i] (= F_{i+2}), low -> high order
    rem = value
    for bit_index in range(len(fibs) - 1, -1, -1):   # greedy: subtract the largest Fibonacci that fits
        if fibs[bit_index] <= rem:
            bits[bit_index] = "1"
            rem -= fibs[bit_index]
    assert rem == 0, f"Zeckendorf decomposition failed for {value}"
    return "".join(bits) + "1"               # append terminator -> unique trailing '11'


def decode(code: str) -> int:
    """Inverse of :func:`encode`: one Fibonacci code (bit-string ending in ``11``) -> the integer."""
    if len(code) < 2 or code[-2:] != "11":
        raise ValueError(f"not a Fibonacci code (must end in '11'): {code!r}")
    zeck = code[:-1]                          # strip the terminator '1'; the rest is d2 d3 ... d_max
    if "11" in zeck:
        # Zeckendorf digits are never consecutive, so an internal "11" is a bit string `encode` can
        # never emit: a corrupted frame is refused here rather than decoded into a plausible symbol.
        raise ValueError(f"not a Fibonacci code (consecutive 1s before the terminator): {code!r}")
    fibs = [1, 2]
    while len(fibs) < len(zeck):
        fibs.append(fibs[-1] + fibs[-2])
    return sum(fibs[bit_index] for bit_index, bit in enumerate(zeck) if bit == "1")


def encode_stream(values: Iterable[int]) -> str:
    """Concatenate the codes; each code's trailing ``11`` frames the next — zero header, self-delimiting."""
    return "".join(encode(value) for value in values)


def decode_stream(bits: str) -> List[int]:
    """Split a concatenated stream at each self-framing ``11`` and decode each code.

    Within a single code the only ``11`` is its terminator (Zeckendorf forbids consecutive 1s), so the
    first ``11`` at/after a code's start is its boundary. A trailing run of bits with no terminator is an
    incomplete final codeword and is dropped (the receiver requests it via the MMR self-heal, out of scope
    here).
    """
    out: List[int] = []
    start = 0
    bit_index = 0
    bit_count = len(bits)
    while bit_index < bit_count:
        if bits[bit_index] == "1" and bit_index + 1 < bit_count and bits[bit_index + 1] == "1":
            out.append(decode(bits[start:bit_index + 2]))
            bit_index += 2
            start = bit_index
        else:
            bit_index += 1
    return out
